07_leukemia_patient/workflow/workflow.sh

```diff
@@ -0,0 +1,232 @@
+}' >> r2.fastq
+
+# for unmapped reads: they don't have the gx tag, so the CB and UB will be different
+
+samtools view -f 4 $STARSOLO_BAM | samtools sort -n | samtools view | grep -v "CB:Z:-" | grep -v "UB:Z:-" | awk -v seen='empty' -v current="another" 'BEGIN{OFS="\t"} {
+ current = $1; # $1 is the read id
+ if (current != seen) {
+  {print "@" current ";"$22";"$23"\n"$10"\n+\n"$11};
+ } seen=current
+}' >> r2.fastq
+
+echo "finished getting read names"
+
+rm header.txt
+
+# deduplicate .fastqs
+
+cat r2.fastq | paste - - - - | sort -k1,1 -t " " | uniq | tr "\t" "\n" > sorted_duplicate_r2.fastq
+
+pigz sorted_duplicate_r2.fastq
+
+rm r2.fastq
+
+echo "R2 file generated"
+
+# run bwa_mem2
+
+mkdir -p $WD/bwa_mem2/genome
+cd $WD/bwa_mem2/genome
+
+# genome generation
+
+wget http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/"$TRANSCRIPTOME".gz
+
+pigz --decompress *gz
+
+grep "ABL1" "$TRANSCRIPTOME"| sed 's/>//g' > ABL1_human.gtf
+grep "BCR" "$TRANSCRIPTOME"| sed 's/>//g' >  BCR_human.gtf
+
+touch ABL1_human.fa
+touch BCR_human.fa
+
+for i in $(cat ABL1_human.gtf); do
+    samtools faidx $TRANSCRIPTOME "$i" >> ABL1_human.fa
+done
+
+for i in $(cat BCR_human.gtf) ; do
+    samtools faidx $TRANSCRIPTOME "$i"  >> BCR_human.fa
+done
+
+cat "$CUSTOM_FA" ABL1_human.fa BCR_human.fa > combined.fa
+cat "$CUSTOM_GTF" ABL1_human.gtf BCR_human.gtf > combined.gtf
+
+# index
+
+~/leukemia_bwamem2/bwa-mem2-2.2.1_x64-linux/bwa-mem2 index -p indexed combined.fa
+
+echo "BWAMEM2 finished"
+
+# run
+
+mkdir -p $WD/bwa_mem2/output
+cd $WD/bwa_mem2/output
+
+~/leukemia_bwamem2/bwa-mem2-2.2.1_x64-linux/bwa-mem2 mem $WD/bwa_mem2/genome/indexed $WD/sorted_duplicate_r2.fastq.gz -t 20 -k 80 -w 12 | samtools view -bS | samtools sort -o bwa_mem2.sorted.bam 
+
+# run STAR fusion --> the genome takes too long to generate (3 days), use premade genome
+
+mkdir -p $WD/star_fusion/output
+cd $WD/star_fusion/output
+
+singularity exec -e -B `pwd` -B $PATH_STAR_FUSION \
+       $PATH_SIMG_FILE \
+       STAR-Fusion \
+       --left_fq $WD/sorted_duplicate_r2.fastq.gz \
+       --genome_lib_dir $STAR_FUSION_GENOME \
+       --output_dir $WD/star_fusion/output > log.txt
+
+echo "STAR_fusion finished"
+
+# adding the CB and UB tag to the bwa mem2 bam file and generating count table
+
+cd $WD/bwa_mem2/output
+
+samtools view -b -F 4 bwa_mem2.sorted.bam > mapped.bam
+
+# removed unmapped reads
+
+samtools view -H mapped.bam > header.txt
+samtools view mapped.bam | awk 'BEGIN {OFS="\t"} {split($1, a, /[;,]/); print a[1], $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, a[2], a[3]}' | cat header.txt - | samtools view -Sbh > annotated_bwa_mem2.sorted.bam
+
+rm header.txt mapped.bam
+
+# UMI deduplication based on CB and UMI and gene name field
+
+echo "# non deduplicated reads bwa mem2"
+
+samtools view annotated_bwa_mem2.sorted.bam | cut -f 1 | wc -l
+
+# bwa mem2 does not require deduplication as the reads are already primary alignments but running anyway
```

Comment:
I don't understand this esntence, do you count only the prim alignment for each read, discarding alt alignments? Hence counting unique and multimappers, but picking only one from the latter? This is tricky, given your genome has bcr and abl wts and fused, I'd expect lots of multimappers. How often do you encounter multimappers?

Reply:
bwa mem2 automatically only reports the primary alignments, while the secondary alignments are stored in the XA tag. Currently, I'm only taking the primary alignments and not taking into account the secondary ones.